Better check for VCS location

lorin opened the pull request:
Identify the likely root directory to pass to the vcs.get_backend_name instead of simply using dist.location. This is necessary to support version control systems such as git and mercurial where the VCS metadata directory only exists off of the root directory.

This is a fix for #713, where "pip freeze" outputs the wrong value for a repository checked out from git if the package puts files in the "src"  directory.

This change is based on @qwcode's comments at https://github.com/pypa/pip/pull/636#issuecomment-7892540.

msabramo commented:
Looks like @qwcode asked the submitter to make some changes and he made them but @qwcode was busy at the time and never got around to reviewing them, but now it doesn't merge cleanly so we'd have to ask @lorin to update if it's still wanted.

lorin commented:
I updated based on the latest develop, but I didn't get a chance to update the new test I added, since the testing style seems to have changed a lot, so I simply dropped it.

The original test I had written depended on a third-party repo ([dcramer/piplint](http://github.com/dcramer/piplint)), so it probably shouldn't be part of the tests anyways.

If there's interest in getting this merged in with a test, could somebody with access to the pypa team create a test repository that conforms to the bug described in #713 so it could be used in the pip tests?

lorin commented:
This has been superseded by #3258.
